src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+        go (pred n) is (mapFreeVars pred n t)
+
+-- ** Tests.
+private
+
+  -- * The tactics works for one or zero hypotheses.
+  data ℝ : ℕ → ℕ → Set where
+    base :
+      ────────────────────────────────
+      ℝ 0 0
+
+    step : ∀ {n m} →
+      ∙ ℝ n m
+      ────────────────────────────────
+      ℝ (suc n) (suc m)
+
+  module _ ⦃ _ : Dec² ℝ ⦄ where
+    unquoteDecl ℝ-base-premises = genPremises ℝ-base-premises (quote ℝ.base)
+    unquoteDecl ℝ-step-premises = genPremises ℝ-step-premises (quote ℝ.step)
+
+    _ : ℝ-base-premises .proj₁
+    _ = tt
+
+    _ : ℝ-base-premises .proj₂ ≡ yes tt
+    _ = refl
+
+    module _ (n m : ℕ) (𝕣 : ℝ n m) where
+      _ : ℝ-step-premises .proj₁
+      _ = 𝕣
+
+      _ : Dec (ℝ n m)
+      _ = ℝ-step-premises .proj₂
+
+  open import Class.Monoid.Core
+  open import Class.Monoid.Instances
+
+  -- * The tactic works for multiple *decidable* hypotheses.
+  data ℚ {A : Set} ⦃ _ : Monoid A ⦄ : A → A → Set where
+    base :
+      ────────────────────────────────
+      ℚ ε ε
+
+    step : ∀ {n m n′ m′} →
+      ∙ ℚ n m
+      ∙ ℚ n′ m′
+      ────────────────────────────────
+      ℚ (n ◇ n′) (m ◇ m′)
+
+  module _ ⦃ _ : ∀ {A} ⦃ _ : Monoid A ⦄ → Dec² (ℚ {A}) ⦄ where
+    unquoteDecl ℚ-base-premises = genPremises ℚ-base-premises (quote ℚ.base)
+    unquoteDecl ℚ-step-premises = genPremises ℚ-step-premises (quote ℚ.step)
+
+    _ : ℚ-base-premises .proj₁
+    _ = tt
+
+    module _ {A} ⦃ _ : Monoid A ⦄ (n m n′ m′ : A) (𝕢ˡ : ℚ n m) (𝕢ʳ : ℚ n′ m′) where
+      _ : ℚ-step-premises .proj₁
+      _ = 𝕢ˡ , 𝕢ʳ
+
+  -- * The tactic omits *undecidable* hypotheses.
+  data ℝ′ : ℕ → ℕ → Set where
+    base :
+      ────────────────────────────────
+      ℝ′ 0 0
+
+    step : ∀ {n m} {T T′ : ℕ → ℕ} →
+      ∙ ℝ′ n m
+      ∙ T ≡ T′
+      ────────────────────────────────
+      ℝ′ (suc n) (suc m)
+
+  module _ ⦃ _ : Dec² ℝ′ ⦄ where
+    unquoteDecl ℝ′-base-premises = genPremises ℝ′-base-premises (quote ℝ′.base)
+    unquoteDecl ℝ′-step-premises = genPremises ℝ′-step-premises (quote ℝ′.step)
+
+    _ : ℝ′-base-premises .proj₁
+    _ = tt
+
+    module _ (n m : ℕ) (𝕣 : ℝ′ n m) where
+      _ : ℝ′-step-premises .proj₁
+      _ = 𝕣
+
+  -- * The tactic works under module contexts.
+  module _ {A} ⦃ _ : Monoid A ⦄ ⦃ _ : Dec² (ℚ {A}) ⦄ ⦃ _ : Dec² ℝ ⦄ where
+
+    data 𝕎 : A → ℕ → Set where
+      base :
+        ────────────────────────────────
+        𝕎 ε 0
+
+      step : ∀ {x y n n′} →
+        ∙ ℝ n n′
+        ∙ ℚ x y
+        ∙ 𝕎 (x ◇ y) n
+        ────────────────────────────────
+        𝕎 (x ◇ y) n′
+
+    unquoteDecl 𝕎-base-premises = genPremises 𝕎-base-premises (quote 𝕎.base)
+
+    _ : 𝕎-base-premises .proj₁
+    _ = tt
+
+    unquoteDecl 𝕎-step-premises = genPremises 𝕎-step-premises (quote 𝕎.step)
+
+    module _ (n n′ : ℕ) (x y : A) (𝕣 : ℝ n n′) (𝕢 : ℚ x y) where
+      _ : 𝕎-step-premises .proj₁
+      _ = 𝕣 , 𝕢
+
+  -- * Irrelevant instance `⦃ Show A ⦄` is not included in the premise type.
```

Comment:
This is crucial if we want to select the *minimal relevant context* w.r.t. the subset of (decidable) hypotheses, since there might be variables that only involve the undecidable fragment or even the conclusion.